streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -2131,4 +2263,11 @@ object ZChannel {
     ): ZChannel[Env1, InErr, InElem, InDone, OutErr, OutElem, OutDone] =
       self.provideSomeEnvironment(_.updateAt(key)(f))
   }
+
+  private case class QRes[A](value: A)
```

Comment:
`QRes` is used to distinguish between user specified type (i.e. the Channel's OutElem) we don't know anything about, and 'control' types like `Cause`.
I guess it could be dropped if the code was migrated to scala3, using union types, the basic idea is to avoid wrapping the stream elements themselves (as they make the vast majority of messages running through the queue) while wrapping the completion/failure message. this way the reader can do a very simple pattern matching to identify the last (success/fail) message and quickly handle the data elements themselves

Reply:
I think @paulpdaniels was referring to having QRes extend `AnyVal` to avoid allocations, like:

```scala
private case class QRes[A](value: A) extends AnyVal
```

Reply:
might be possible, checking